code/render/batching.h

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) Freespace Open 2015.  All rights reserved.
+ *
+ * All source code herein is the property of Freespace Open. You may not sell 
+ * or otherwise commercially exploit the source or things you created based on the 
+ * source.
+ *
+*/
+
+#include "globalincs/pstypes.h"
+#include "graphics/2d.h"
+#include "graphics/material.h"
+
+struct batch_vertex {
+	vec3d position;
+	uv_pair tex_coord;
+	ubyte r, g, b, a;
+	float radius;
+	vec3d uvec;
+};
+
+struct batch_info {
+	enum material_type {
+		FLAT_EMISSIVE,
+		VOLUME_EMISSIVE,
+		DISTORTION,
+		NUM_RENDER_TYPES
+	};
+
+	material_type mat_type;
+	int texture;
+	primitive_type prim_type;
+	bool thruster;	// only used by distortion
+
+	batch_info(): mat_type(FLAT_EMISSIVE), texture(-1), prim_type(PRIM_TYPE_TRIS), thruster(false) {}
+	batch_info(material_type mat, int tex, primitive_type prim, bool thrust): mat_type(mat), texture(tex), prim_type(prim), thruster(thrust) {}
+
+	bool operator < (const batch_info& batch) const {
+		if ( mat_type != batch.mat_type ) {
+			return mat_type < batch.mat_type;
+		}
+
+		if ( texture != batch.texture ) {
+			return texture < batch.texture;
+		}
+
+		if ( prim_type != batch.prim_type ) {
+			return prim_type < batch.prim_type;
+		}
+
+		return thruster != batch.thruster;
+	}
+};
+
+struct batch_buffer_key {
+	uint Vertex_mask;
+	primitive_type Prim_type;
+
+	batch_buffer_key(): Vertex_mask(0), Prim_type(PRIM_TYPE_TRIS) {}
+	batch_buffer_key(uint vertex_mask, primitive_type prim_type): Vertex_mask(vertex_mask), Prim_type(prim_type) {}
+
+	bool operator < (const batch_buffer_key& key) const {
+		if ( Vertex_mask != key.Vertex_mask ) {
+			return Vertex_mask < key.Vertex_mask;
+		}
+
+		return Prim_type < key.Prim_type;
+	}
+};
+
+class primitive_batch
+{
+	batch_info render_info;
+	SCP_vector<batch_vertex> Vertices;
+
+public:
+	primitive_batch() : render_info() {}
+	primitive_batch(batch_info info): render_info(info) {}
+
+	batch_info &get_render_info() { return render_info; }
+
+	void add_triangle(batch_vertex* v0, batch_vertex* v1, batch_vertex* v2);
+	void add_point_sprite(batch_vertex *p);
+
+	int load_buffer(batch_vertex* buffer, int n_verts);
+
+	size_t num_verts() { return Vertices.size();  }
+
+	void clear();
+};
+
+struct primitive_batch_item {
+	batch_info batch_item_info;
+	int offset;
+	int n_verts;
```

Comment:
These two fields should probably use `size_t` or `ptrdiff_t` since they are dependent on the pointer size.